x-pack/legacy/plugins/apm/server/lib/service_map/get_service_map_from_trace_ids.ts

```diff
@@ -0,0 +1,279 @@
+                  'processor.event',
+                  'span.type',
+                  'span.subtype',
+                  'agent.name'
+                };
+                state.fieldsToCopy = fieldsToCopy;`
+            },
+            map_script: {
+              lang: 'painless',
+              source: `def id;
+                if (!doc['span.id'].empty) {  
+                  id = doc['span.id'].value;
+                } else {
+                  id = doc['transaction.id'].value;
+                }
+
+                def copy = new HashMap();
+                copy.id = id;
+
+                for(key in state.fieldsToCopy) {
+                  if (!doc[key].empty) {
+                    copy[key] = doc[key].value;
+                  }
+                }
+
+                state.eventsById[id] = copy`
+            },
+            combine_script: {
+              lang: 'painless',
+              source: `return state.eventsById;`
+            },
+            reduce_script: {
+              lang: 'painless',
+              source: `
+              def getDestination ( def event ) {
+                def destination = new HashMap();
+                destination['destination.address'] = event['destination.address'];
+                destination['span.type'] = event['span.type'];
+                destination['span.subtype'] = event['span.subtype'];
+                return destination;
+              }
+              
+              def processAndReturnEvent(def context, def eventId) {
+                if (context.processedEvents[eventId] != null) {
+                  return context.processedEvents[eventId];
+                }
+                
+                def event = context.eventsById[eventId];
+
+                if (event == null) {
+                  return null;
+                }
+
+                def service = new HashMap();
+                service['service.name'] = event['service.name'];
+                service['service.environment'] = event['service.environment'];
+                service['agent.name'] = event['agent.name'];
+
+                def basePath = new ArrayList();
+
+                def parentId = event['parent.id'];
+                def parent;
+
+                if (parentId != null && parentId != event['id']) {
+                  parent = processAndReturnEvent(context, parentId);
+                  if (parent != null) {
+                    /* copy the path from the parent */
+                    basePath.addAll(parent.path);
+                    /* flag parent path for removal, as it has children */
+                    context.locationsToRemove.add(parent.path);
+
+                    /* if the parent has 'destination.address' set, and the service is different,
+                    we've discovered a service */
+
+                    if (parent['destination.address'] != null
+                      && parent['destination.address'] != ""
+                      && parent['span.type'] == 'external'
```

Comment:
we should add `span.type == 'messaging'` here, I think

Reply:
so `&& parent['span.type'] == 'external' || parent['span.type'] == 'messaging'`? is this for gRPC?

Reply:
just messaging systems in general, not sure what kind of practical implementations there are